### PR TITLE
E2E testing

### LINKED ISSUE #317: proposal: introduce an e2e test suite
Refer to the e2e test in osmosis and Gaia. I propose we should introduce an e2e integration test suite that allows us to test chain functionality  :
  - IBC transfers
  - Chain upgrade
  - Interact with Wasm Contract 
 
We can refer to Osmosis and Gaia  : 
  - https://github.com/cosmos/gaia/tree/main/tests/e2e
  - https://github.com/osmosis-labs/osmosis/tree/main/tests/e2e

### PR DESCRIPTION
**Pull request by vuong177**
This PR is for CI about the new Juno e2e test.

Close #317

### CODE REVIEW

File: tests/e2e/containers/config.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+const (
+	// Current Git branch juno repo/version. It is meant to be built locally.
+	// It is used when skipping upgrade by setting JUNO_E2E_SKIP_UPGRADE to true).
+	// This image should be pre-built with `make docker-build-debug` either in CI or locally.
+	CurrentBranchRepository = "juno"
+	CurrentBranchTag        = "debug"
+	// Pre-upgrade juno repo/tag to pull.
+	// It should be uploaded to Docker Hub. JUNO_E2E_SKIP_UPGRADE should be unset
+	// for this functionality to be used.
+	previousVersionRepository = "ducnt87/juno"
```
Comment (the-frey) [inline tests/e2e/containers/config.go:26]: is this right?

Comment (vuong177) [reply tests/e2e/containers/config.go:26]: Notional team can't access to CosmosContract DockerHub, so we need to create one by personal account. I think better if we can automate creating this and uploading it by the official DockerHub account.

Comment (anhductn2001) [reply tests/e2e/containers/config.go:26]: Hi, This is a temporary solution for pulling image from my Dockerhub. I hope to be able to push it to an official Juno account in the future.

Comment (anhductn2001) [reply tests/e2e/containers/config.go:26]: And I wills setup a bot that automatically creates images on DockerHub every time a new version is available. To do this I will need an official Dockerhub of Juno.

Comment (the-frey) [reply tests/e2e/containers/config.go:26]: If you are in the notional team for Juno you should be able to make branches on this repo, then you can use the github images docker repository, if that's what you mean?

Comment (anhductn2001) [reply tests/e2e/containers/config.go:26]: Right!!! But I don't have permission to make a branch on Juno right now, I will ask Vuong to create it.

Comment (vuong177) [reply tests/e2e/containers/config.go]: @the-frey 
seem I need more permission for push to ghcr. It's my log :
```
The push refers to repository [[ghcr.io/cosmoscontracts/juno-e2e](http://ghcr.io/cosmoscontracts/juno-e2e?fbclid=IwAR1NXhkAZvYN6JgMM44mF41fQ-I-qJMGDP0_vllLNwTNUAQ9peAv1BMeFIw)]
8a2d6dd588ba: Preparing 
36a41810fd3a: Preparing 
34d5ebaa5410: Preparing 
denied: permission_denied: The token provided does not match expected scopes.
```

Comment (anhductn2001) [reply tests/e2e/containers/config.go]: @the-frey I have uploaded the image to ghcr.io/cosmoscontracts/juno...Can you review it? Many thanks!!

**Review by faddat — DISMISSED** [review]

**Review by faddat — APPROVED** [review]

**Review by catShaark — APPROVED** [review]

**Comment by vuong177** [comment]
@the-frey, could you take a look on this PR ? This PR is a bit big so need your review !

**Comment by the-frey** [comment]
Okay, this is a big PR. Will take me a little while to digest haha.

Is there anything in here you're particularly worried about or think is spicy?

**Comment by faddat** [comment]
oh crap

**Comment by the-frey** [comment]
🤦 I will look later and shout if there is something of concern.
Hopefully as it is just testing scaffolding there is nothing that could be an issue, but it is a big PR.

**Pull request merged**